lib/bridge.js

```diff
@@ -788,66 +791,97 @@ Bridge.prototype._onEvent = function(event) {
             const isUpgradeInvite = this._roomUpgradeHandler.onInvite(event);
             if (isUpgradeInvite &&
                 this.opts.roomUpgradeOpts.consumeEvent) {
-                return Promise.resolve();
+                return null;
             }
         }
     }
 
-    var request = this._requestFactory.newRequest({ data: event });
-    var context = new BridgeContext({
-        sender: event.user_id,
-        target: event.state_key,
-        room: event.room_id
-    });
-    var data = {
-        request: request,
-        context: context
-    };
+    const request = this._requestFactory.newRequest({ data: event });
+    const contextReady = this._getBridgeContext(event);
+    const dataReady = contextReady.then(context => ({ request, context }));
 
-    var promise;
-    if (this.opts.disableContext) {
-        promise = Promise.resolve();
-        data.context = null;
-    }
-    else {
-        promise = context.get(this._roomStore, this._userStore);
-    }
+    const dataReadyLimited = this._limited(dataReady, request);
 
-    if (this.opts.queue.type === "none") { // consume as soon as we have context
-        promise.done(() => {
-            this._onConsume(null, data);
-        }, (err) => {
-            this._onConsume(err);
-        });
-        return request.getPromise();
-    }
+    this._queue.push(event, dataReadyLimited);
+    this._queue.consume();
 
+    return (
+        request
+        .getPromise()
+        .catch(EventNotHandledError, e => this._handleEventError(event, e))
+    );
+};
+
+/**
+ * Returns a promise that is like `promise` but adheres to all necessary limits.
```

Comment:
This needs to be less vague. What are "all necessary limits"?

Reply:
Hope its clearer now.